File: application/backend-credit-card-enrollment/backend-java/src/main/java/cloud/ambar/common/serializedevent/Serializer.java
package cloud.ambar.common.serializedevent;

import cloud.ambar.common.event.Event;
import cloud.ambar.creditcard.enrollment.event.EnrollmentAccepted;
import cloud.ambar.creditcard.enrollment.event.EnrollmentDeclined;
import cloud.ambar.creditcard.enrollment.event.EnrollmentRequested;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;

@Service
@RequiredArgsConstructor
public class Serializer {
    private final ObjectMapper objectMapper = new ObjectMapper();

    public SerializedEvent serialize(Event event) {
        return SerializedEvent.builder()
                .eventId(event.getEventId())
                .aggregateId(event.getAggregateId())
                .aggregateVersion(event.getAggregateVersion())
                .correlationId(event.getCorrelationId())
                .causationId(event.getCausationId())
                .recordedOn(formatInstant(event.getRecordedOn()))
                .eventName(determineEventName(event))
                .jsonPayload(createJsonPayload(event))
                .jsonMetadata("{}")
                .build();
    }

    private String determineEventName(Event event) {
        return switch (event) {
            case EnrollmentRequested _event -> "CreditCard_Enrollment_EnrollmentRequested";
            case EnrollmentAccepted _event -> "CreditCard_Enrollment_EnrollmentAccepted";
            case EnrollmentDeclined _event -> "CreditCard_Enrollment_EnrollmentDeclined";
            case null -> throw new RuntimeException("Event is null");
            default -> {
                throw new RuntimeException("Unknown event type: " + event.getClass().getName());
            }
        };
    }

    private String createJsonPayload(Event event) {
        try {
            ObjectNode jsonPayload = objectMapper.createObjectNode();

            if (event instanceof EnrollmentRequested enrollmentRequested) {
                jsonPayload.put("annualIncomeInCents", enrollmentRequested.getAnnualIncomeInCents());
                jsonPayload.put("productId", enrollmentRequested.getProductId());
                jsonPayload.put("userId", enrollmentRequested.getUserId());
            } else if (event instanceof EnrollmentAccepted enrollmentAccepted) {
                jsonPayload.put("reasonCode", enrollmentAccepted.getReasonCode());
                jsonPayload.put("reasonDescription", enrollmentAccepted.getReasonDescription());
            }  else if (event instanceof EnrollmentDeclined enrollmentDeclined) {
                jsonPayload.put("reasonCode", enrollmentDeclined.getReasonCode());
                jsonPayload.put("reasonDescription", enrollmentDeclined.getReasonDescription());
            } else {
                return "{}";
            }

            return objectMapper.writeValueAsString(jsonPayload);
        } catch (Exception e) {
            throw new RuntimeException("Failed to serialize event payload", e);
        }
    }

    private String formatInstant(Instant instant) {
        ZonedDateTime zdt = instant.atZone(ZoneId.of("UTC"));
        return DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSSSSS z").format(zdt);
    }
}

File: application/backend-credit-card-enrollment/backend-c#/CreditCardEnrollment/Common/Events/Serializer.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using CreditCardEnrollment.Domain.Enrollment.Events;
using CreditCardEnrollment.Domain.Product.Events;

namespace CreditCardEnrollment.Common.Events;

public interface ISerializer
{
    SerializedEvent Serialize(Event @event);
}

public class Serializer : ISerializer
{
    public SerializedEvent Serialize(Event @event)
    {
        return new SerializedEvent
        {
            EventId = @event.EventId,
            AggregateId = @event.AggregateId,
            AggregateVersion = @event.AggregateVersion,
            CorrelationId = @event.CorrelationId,
            CausationId = @event.CausationId,
            RecordedOn = FormatDateTime(@event.RecordedOn),
            EventName = DetermineEventName(@event),
            JsonPayload = CreateJsonPayload(@event),
            JsonMetadata = "{}"
        };
    }

    private static string DetermineEventName(Event @event) => @event switch
    {
        EnrollmentRequested => "CreditCard_Enrollment_EnrollmentRequested",
        EnrollmentAccepted => "CreditCard_Enrollment_EnrollmentAccepted",
        EnrollmentDeclined => "CreditCard_Enrollment_EnrollmentDeclined",
        ProductActivated => "CreditCard_Product_ProductActivated",
        ProductDeactivated => "CreditCard_Product_ProductDeactivated",
        ProductDefined => "CreditCard_Product_ProductDefined",
        _ => throw new ArgumentException($"Unknown event type: {@event.GetType().Name}")
    };

    private static string CreateJsonPayload(Event @event)
    {
        var jsonObject = new JsonObject();

        switch (@event)
        {
            case EnrollmentRequested enrollmentRequested:
                jsonObject.Add("annualIncomeInCents", enrollmentRequested.AnnualIncomeInCents);
                jsonObject.Add("productId", enrollmentRequested.ProductId);
                jsonObject.Add("userId", enrollmentRequested.UserId);
                break;

            case EnrollmentAccepted enrollmentAccepted:
                jsonObject.Add("reasonCode", enrollmentAccepted.ReasonCode);
                jsonObject.Add("reasonDescription", enrollmentAccepted.ReasonDescription);
                jsonObject.Add("userId", enrollmentAccepted.UserId);
                jsonObject.Add("productId", enrollmentAccepted.ProductId);
                break;

            case EnrollmentDeclined enrollmentDeclined:
                jsonObject.Add("reasonCode", enrollmentDeclined.ReasonCode);
                jsonObject.Add("reasonDescription", enrollmentDeclined.ReasonDescription);
                jsonObject.Add("userId", enrollmentDeclined.UserId);
                jsonObject.Add("productId", enrollmentDeclined.ProductId);
                break;

            case ProductDefined productDefined:
                jsonObject.Add("name", productDefined.Name);
                jsonObject.Add("interestInBasisPoints", productDefined.InterestInBasisPoints);
                jsonObject.Add("annualFeeInCents", productDefined.AnnualFeeInCents);
                jsonObject.Add("paymentCycle", productDefined.PaymentCycle);
                jsonObject.Add("creditLimitInCents", productDefined.CreditLimitInCents);
                jsonObject.Add("maxBalanceTransferAllowedInCents", productDefined.MaxBalanceTransferAllowedInCents);
                jsonObject.Add("reward", productDefined.Reward);
                jsonObject.Add("cardBackgroundHex", productDefined.CardBackgroundHex);
                break;
        }

        return jsonObject.ToJsonString();
    }

    private static string FormatDateTime(DateTime dateTime)
    {
        // Format to match Java's "yyyy-MM-dd HH:mm:ss.SSSSSS z" pattern
        return dateTime.ToUniversalTime().ToString("yyyy-MM-dd HH:mm:ss.ffffff UTC");
    }
}
